v03_pipeline/lib/misc/lookup.py

```diff
@@ -0,0 +1,234 @@
+import hail as hl
+
+from v03_pipeline.lib.model import DatasetType
+
+
+def compute_callset_lookup_ht(
+    dataset_type: DatasetType,
+    mt: hl.MatrixTable,
+    project_guid: str,
+) -> hl.Table:
+    sample_id_to_family_guid = hl.dict(
+        {
+            s: family_guid
+            for family_guid, sample_ids in hl.eval(mt.family_samples).items()
+            for s in sample_ids
+        },
+    )
+    ht = mt.select_rows(
+        project_stats=hl.array(
+            [
+                hl.agg.collect(
+                    hl.Struct(
+                        family_guid=sample_id_to_family_guid[mt.s],
+                        **mt.entry,
+                    ),
+                )
+                .group_by(lambda fs: fs.family_guid)
+                .values()
+                .map(
+                    lambda fs: hl.Struct(
+                        family_guid=fs[0].family_guid,
+                        **{
+                            field_name: hl.len(fs.filter(filter_fn))
+                            for field_name, filter_fn in dataset_type.lookup_table_fields_and_genotype_filter_fns.items()
+                        },
+                    ),
+                ),
+            ],
+        ),
+    ).rows()
+    return globalize_ids(ht, project_guid)
+
+
+def globalize_ids(ht: hl.Table, project_guid: str) -> hl.Table:
+    row = ht.take(1)
+    has_project_stats = row and len(row[0].project_stats) > 0
+    ht = ht.annotate_globals(
+        project_guids=[project_guid],
+        project_families=(
+            {project_guid: [fs.family_guid for fs in ps] for ps in row[0].project_stats}
+            if has_project_stats
+            else hl.empty_dict(hl.tstr, hl.tarray(hl.tstr))
+        ),
+    )
+    return ht.annotate(
+        project_stats=ht.project_stats.map(
+            lambda ps: ps.map(lambda fs: fs.drop('family_guid')),
+        ),
+    )
+
+
+def remove_new_callset_family_guids(
+    ht: hl.Table,
+    project_guid: str,
+    family_guids: hl.SetExpression,
+) -> hl.Table:
+    if project_guid not in hl.eval(ht.globals.project_families):
+        return ht
+    project_i = ht.project_guids.index(project_guid)
+    family_indexes_to_keep = hl.array(
+        hl.enumerate(ht.globals.project_families[project_guid])
+        .filter(lambda item: ~family_guids.contains(item[1]))
+        .map(lambda item: item[0]),
+    )
+    ht = ht.annotate(
+        project_stats=(
+            hl.enumerate(ht.project_stats).starmap(
+                lambda i, fs: (
+                    hl.if_else(
+                        i != project_i,
```

Comment:
probably out of scope, but since we are often running the pipeline on multiple projects at once it feels kind of inefficient that we need to run this operation once per project, it feels like it would be more efficient to run it for all the projects at once

Reply:
👍 should be a straightforward change I think.  Would save I/O at least!